src/script.c

```diff
@@ -0,0 +1,434 @@
+    while (RunScriptCommand(&sScriptContext2) == TRUE);
+}
+
+u8 *mapheader_get_tagged_pointer(u8 tag)
+{
+    u8 *mapScripts = gMapHeader.mapScripts;
+
+    if (!mapScripts)
+        return NULL;
+
+    while (1)
+    {
+        if (!*mapScripts)
+            return NULL;
+        if (*mapScripts == tag)
+        {
+            mapScripts++;
+            return (u8 *)(mapScripts[0] + (mapScripts[1] << 8) + (mapScripts[2] << 16) + (mapScripts[3] << 24));
+        }
+        mapScripts += 5;
+    }
+}
+
+void mapheader_run_script_by_tag(u8 tag)
+{
+    u8 *ptr = mapheader_get_tagged_pointer(tag);
+    if (ptr)
+        ScriptContext2_RunNewScript(ptr);
+}
+
+u8 *mapheader_get_first_match_from_tagged_ptr_list(u8 tag)
+{
+    u8 *ptr = mapheader_get_tagged_pointer(tag);
+
+    if (!ptr)
+        return NULL;
+
+    while (1)
+    {
+        u16 varIndex1;
+        u16 varIndex2;
+        varIndex1 = ptr[0] | (ptr[1] << 8);
+        if (!varIndex1)
+            return NULL;
+        ptr += 2;
+        varIndex2 = ptr[0] | (ptr[1] << 8);
+        ptr += 2;
+        if (VarGet(varIndex1) == VarGet(varIndex2))
+            return (u8 *)(ptr[0] + (ptr[1] << 8) + (ptr[2] << 16) + (ptr[3] << 24));
+        ptr += 4;
+    }
+}
+
+void mapheader_run_script_with_tag_x1(void)
+{
+    mapheader_run_script_by_tag(1);
+}
+
+void mapheader_run_script_with_tag_x3(void)
+{
+    mapheader_run_script_by_tag(3);
+}
+
+void mapheader_run_script_with_tag_x5(void)
+{
+    mapheader_run_script_by_tag(5);
+}
+
+void mapheader_run_script_with_tag_x7(void)
+{
+    mapheader_run_script_by_tag(7);
+}
+
+void mapheader_run_script_with_tag_x6(void)
+{
+    mapheader_run_script_by_tag(6);
+}
+
+bool8 mapheader_run_first_tag2_script_list_match(void)
+{
+    u8 *ptr = mapheader_get_first_match_from_tagged_ptr_list(2);
+
+    if (!ptr)
+        return 0;
+
+    ScriptContext1_SetupScript(ptr);
+    return 1;
+}
+
+void mapheader_run_first_tag4_script_list_match(void)
+{
+    u8 *ptr = mapheader_get_first_match_from_tagged_ptr_list(4);
+    if (ptr)
+        ScriptContext2_RunNewScript(ptr);
+}
+
+u32 CalculateRamScriptChecksum(void)
+{
+    return CalcCRC16WithTable((u8*)(&gSaveBlock1Ptr->ramScript.data), sizeof(gSaveBlock1Ptr->ramScript.data));
+}
+
+void ClearRamScript(void)
+{
+    CpuFill32(0, &gSaveBlock1Ptr->ramScript, sizeof(struct RamScript));
+}
+
+bool8 InitRamScript(u8 *script, u16 scriptSize, u8 mapGroup, u8 mapNum, u8 objectId)
+{
+    struct RamScriptData *scriptData = &gSaveBlock1Ptr->ramScript.data;
+
+    ClearRamScript();
+
+    if (scriptSize > sizeof(scriptData->script))
+        return FALSE;
+
+    scriptData->magic = RAM_SCRIPT_MAGIC;
+    scriptData->mapGroup = mapGroup;
+    scriptData->mapNum = mapNum;
+    scriptData->objectId = objectId;
+    memcpy(scriptData->script, script, scriptSize);
+    gSaveBlock1Ptr->ramScript.checksum = CalculateRamScriptChecksum();
+    return TRUE;
+}
+
+u8 *GetRamScript(u8 objectId, u8 *script)
```

Comment:
Did this not work with the same if structure from pokeruby?

Reply:
Nope, the ruby one is a 'fake-match'. :P

Reply:
Yeah, I tried porting the Ruby one (which looks better) here, and it had stupid register differences. Oh well.